src/Package/Impl/Repl/VsRInteractiveWindowComponentContainerFactory.cs

```diff
@@ -14,39 +15,67 @@
+            Lazy<IVsInteractiveWindowFactory> vsInteractiveWindowFactory,
+            IContentTypeRegistryService contentTypeRegistryService) {
             _vsInteractiveWindowFactoryLazy = vsInteractiveWindowFactory;
             _contentTypeRegistryService = contentTypeRegistryService;
+
         }
 
-        public IInteractiveWindowVisualComponent Create(int instanceId, IInteractiveEvaluator evaluator) {
+        public IInteractiveWindowVisualComponent Create(int instanceId, IInteractiveEvaluator evaluator, IRInteractiveWorkflow interactiveWorkflow) {
             VsAppShell.Current.AssertIsOnMainThread();
 
+            _interactiveWorkflow = interactiveWorkflow;
+            _interactiveWorkflow.RSessions.BrokerChanged += OnBrokerChanged;
+
             var vsWindow = _vsInteractiveWindowFactoryLazy.Value.Create(RGuidList.ReplInteractiveWindowProviderGuid, instanceId, Resources.ReplWindowName, evaluator);
             var contentType = _contentTypeRegistryService.GetContentType(RContentTypeDefinition.ContentType);
             vsWindow.SetLanguage(RGuidList.RLanguageServiceGuid, contentType);
 
-            var toolWindow = (ToolWindowPane) vsWindow;
-            var frame = (IVsWindowFrame)toolWindow.Frame;
-            frame.SetProperty((int)__VSFPROPID4.VSFPROPID_TabImage, Resources.ReplWindowIcon);
+            var toolWindow = (ToolWindowPane)vsWindow;
+            _frame = (IVsWindowFrame)toolWindow.Frame;
+            _frame.SetProperty((int)__VSFPROPID4.VSFPROPID_TabImage, Resources.ReplWindowIcon);
             // TODO: figure out why REPL window doesn't get 'force create' flag set
             // For now, set it forcibly when window is shown
             object value;
-            frame.GetProperty((int)__VSFPROPID.VSFPROPID_CreateToolWinFlags, out value);
-            frame.SetProperty((int)__VSFPROPID.VSFPROPID_CreateToolWinFlags, (int)value | (int)__VSCREATETOOLWIN.CTW_fForceCreate);
+            _frame.GetProperty((int)__VSFPROPID.VSFPROPID_CreateToolWinFlags, out value);
+            _frame.SetProperty((int)__VSFPROPID.VSFPROPID_CreateToolWinFlags, (int)value | (int)__VSCREATETOOLWIN.CTW_fForceCreate);
+
+            UpdateWindowTitle();
 
             var componentContainer = new VisualComponentToolWindowAdapter<IInteractiveWindowVisualComponent>(toolWindow);
             var component = new RInteractiveWindowVisualComponent(vsWindow.InteractiveWindow, componentContainer);
             componentContainer.Component = component;
             return component;
         }
+
+        private void OnBrokerChanged(object sender, EventArgs e) {
+            UpdateWindowTitle();
+        }
+
+        private void UpdateWindowTitle() {
+            var broker = _interactiveWorkflow.RSessions.Broker;
+            string title;
+            if (broker != null) {
+                title = broker.IsRemote
+                    ? Invariant($"{Resources.ReplWindowName} - {broker.Name} ({broker.Uri.ToString().TrimTrailingSlash()})")
```

Comment:
It is not something unseen before... 

![image](https://cloud.githubusercontent.com/assets/12820357/18486560/6ca5cbe2-79a6-11e6-9699-95b651613686.png)